### ISSUE TITLE
Scanlines mode options or glsl shader support?

### ISSUE THREAD

**Issue by RetroD** [midwan/amiberry#425]
Would it be possible at all to add some options for scanlines such as opacity (something like 25, 50, 75, 100%) and a filter to soften the image a bit? Even better, can glsl shader support be added so that we can use the same shader library from libretro? (all to be saved globally in the amiberry.conf of course :P ) 

Don't know about the feasibility for such requests, and I understand is not a priority by any means, but I thought I'd throw it out here to see what you guys think :)

**Comment by midwan** [comment]
Yes, I was planning to look into this at some point in the future... :)

**Comment by mrfomt** [comment]
This would be great if you could add to Amiberry!

**Comment by sphaero** [comment]
Any pointers where to start? I know uae4arm-rpi has some shader support. I'm not even sure yet if Amiberry is even using OpenGLES. :thinking:

**Comment by sphaero** [comment]
FWIW, I've just build amiberry from source with USE_OPENGL=ON. I had to change CMakeLists.txt
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 74a8d5ab..04468ae2 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -38,7 +38,7 @@ set(PROJECT_COMPANY_NAMESPACE "com.blitterstudio")  # Reverse domain name notati
 include(GNUInstallDirs)
 
 include(cmake/StandardProjectSettings.cmake)
-include(cmake/Dependencies.cmake)
 include(cmake/SourceFiles.cmake)
+include(cmake/Dependencies.cmake)
```

It builds fine but when I run it I get a few first frames from kickstart booting but it then remains black. When I kill it spits `Vertex Shader Error: Vertex Shader Error: Vertex Shader Error:`

Still quite experimental I guess?

**Comment by midwan** [comment]
Yeah, sorry - this is not implemented completely yet. There are some bugs which I haven't had a chance to fix there.

**Comment by sphaero** [comment]
No worries, I can only salute your work. It seems I get nostalgic every time around christmas and start firing up amiga stuff. It's just too little time to acquaint myself with the source to do something :disappointed:

**Comment by filevans** [comment]
> Yeah, sorry - this is not implemented completely yet. There are some bugs which I haven't had a chance to fix there.

hello is there a way to get any CRT scanline shaders on amiberry on the Pi now? 
thanks

**Comment by midwan** [comment]
@filevans 
No, that's not implemented yet. I'm hoping it will work eventually, but I haven't had enough time to learn the details of why my OpenGL implementation is broken, yet.

**Comment by filevans** [comment]
@midwan  I see, please could it be looked at? As scanlines are really needed to look authentic how an Amiga used to look. Not sure if vulkan is possible instead which uses GLSL?

**Comment by midwan** [comment]
It will be looked at, at some point. :)

**Comment by simpsond** [comment]
@midwan pointed out that Auto Crop doesn't work with my changes above (Thank you).  This is because OpenGL needs a packed pixel buffer.

I'm playing with adding this function:
```
static uae_u8* create_packed_pixel_buffer(const SDL_Surface* src,
	const SDL_Rect& crop, int x_padding, int y_padding, SDL_Rect& out_buffer_rect)
{
	if (!src)
	{
		out_buffer_rect = { 0, 0, 0, 0 };
		return nullptr;
	}

	SDL_Rect desired_crop;
	desired_crop.x = crop.x - x_padding;
	desired_crop.y = crop.y - y_padding;
	desired_crop.w = crop.w + (2 * x_padding);
	desired_crop.h = crop.h + (2 * y_padding);

	const SDL_Rect src_bounds = { 0, 0, src->w, src->h };
	SDL_Rect final_crop;
	if (!SDL_IntersectRect(&desired_crop, &src_bounds, &final_crop))
	{
		out_buffer_rect = { 0, 0, 0, 0 };
		return nullptr;
	}

	const int bytes_per_pixel = src->format->BytesPerPixel;
	const int buffer_row_bytes = final_crop.w * bytes_per_pixel;
	const size_t buffer_size = buffer_row_bytes * final_crop.h;

	if (buffer_size == 0)
	{
	   out_buffer_rect = { 0, 0, 0, 0 };
	   return nullptr;
	}
	uae_u8* packed_buffer = new uae_u8[buffer_size];

	const uae_u8* src_row_start = static_cast<const uae_u8*>(src->pixels)
							  + final_crop.y * src->pitch
							  + final_crop.x * bytes_per_pixel;
	uae_u8* dst_row_start = packed_buffer;

	for (int y = 0; y < final_crop.h; ++y)
	{
		memcpy(dst_row_start, src_row_start, buffer_row_bytes);
		src_row_start += src->pitch;
		dst_row_start += buffer_row_bytes;
	}

	out_buffer_rect = final_crop;
	return packed_buffer;
}
```

You'll notice the function takes a padding adjustment that can be used to compensate for cropping done within the shader.
```
SDL_Rect corrected_crop_rect;
	auto packed_pixel_buffer = create_packed_pixel_buffer(amiga_surface, crop_rect, 8, 40, corrected_crop_rect);
	crtemu_present(crtemu_tv, time, reinterpret_cast<const CRTEMU_U32*>(packed_pixel_buffer),
	  corrected_crop_rect.w, corrected_crop_rect.h, 0xffffffff, 0x000000);
	free(packed_pixel_buffer);
```

The problem is that the TV emulation shader is also simulating TV overscan, which is what causes the cropping.

 The tsample function inside the shader string:

   `tc = tc * vec2(1.025, 0.92) + vec2(-0.0125, 0.04);`

So we get a double crop.

My create_packed_pixel_buffer() function attempts to correct for this, but the alignment is wonky.

TV Shader (create_packed_pixel_buffer called with X:8 Y:40):
<img width="747" height="603" alt="Image" src="https://github.com/user-attachments/assets/f2a0f56a-f5db-49dd-b914-cb2b0376aa91" />

PC Shader(create_packed_pixel_buffer called with X:0 Y:0):

<img width="749" height="602" alt="Image" src="https://github.com/user-attachments/assets/ff0da695-438b-4365-99e5-5edcc441d5a0" />

Lite Shader (create_packed_pixel_buffer called with X:0 Y:0)
<img width="759" height="608" alt="Image" src="https://github.com/user-attachments/assets/2feed21d-2b94-4a22-a46a-76ca81fe7188" />

I super hate the idea of needing to hard code padding offsets in code to compensate for shaders. Any thoughts?
